src/wallet/coinselection.cpp

```diff
@@ -359,6 +367,9 @@ util::Result<SelectionResult> KnapsackSolver(std::vector<OutputGroup>& groups, c
     if (lowest_larger &&
         ((nBest != nTargetValue && nBest < nTargetValue + change_target) || lowest_larger->GetSelectionAmount() <= nBest)) {
         result.AddInput(*lowest_larger);
+
```

Comment:
I would suggest that you pass the `max_weight` as an additional parameter to `ApproximateBestSubset(…)` and keep track of the current weight (e.g. `nTotalWeight`) in parallel to `nTotal` and only designate a new `nBest` and `vfBest` if it’s not just better but also a permitted weight:

```diff
-                        if (nTotal < nBest)
+                        if (nTotal < nBest && nTotalWeight <= max_weight)                        
```

for extra credit you could also deselect not just after exceeding the target but also when you exceed max_weight

Reply:
did the minimal suggested work to give knapsack a fighting chance